clientconn.go

```diff
@@ -1100,11 +1104,18 @@ func (ac *addrConn) createTransport(backoffNum int, addr resolver.Address, copts
 
 		// TODO(deklerk): optimization; does anyone else actually use this lock? maybe we can just remove it for this scope
 		ac.mu.Lock()
-		ac.successfulHandshake = true
-		ac.backoffDeadline = time.Time{}
-		ac.connectDeadline = time.Time{}
-		ac.addrIdx = 0
-		ac.backoffIdx = 0
+
+		prefaceMu.Lock()
+		serverPrefaceReceived = true
+		if clientPrefaceWrote {
```

Comment:
Instead of the extra synchronization and two new locals, could we split ac.successfulHandshake in two, one for client and one for server, and where it's checked in nextAddr, we check for the two &&ed together instead?  Would that be simpler possibly?

Reply:
Thought about doing that, but unfortunately we have that logic about "very first connection attempt" that's tied to that. 🤢 

This seemed simplest until we sort out that whole mess. WDYT?